File: 0662-maximum-width-of-binary-tree/0662-maximum-width-of-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var widthOfBinaryTree = function(root) {
    var maxlength=0;
    const queue =[]
    queue.push({index:0, Node:root})
    while(queue.length){
        let length = queue.length;
        let diff= queue[length-1].index- queue[0].index +1;
        
        maxlength= Math.max(diff,maxlength )
         let min = queue[0].index;
        for(let i=0;i<length;i++){
           
           
            let temp = queue[0];
            queue.shift()
            if (temp.Node.left!=null){
                queue.push({index: 2*(temp.index-min)+1, Node: temp.Node.left})

            }
            if (temp.Node.right!=null){
                queue.push({index: 2*(temp.index-min)+2, Node: temp.Node.right})
            }
        }


    }
    return maxlength;
    
};

File: 0662-maximum-width-of-binary-tree/0662-maximum-width-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
class Solution {
public:
    int widthOfBinaryTree(TreeNode* root) {
        int maxwidth = 1;

        queue<pair<TreeNode*, int>> q;
        q.push({root, 0});
        while (q.size() > 0) {
            int qsize = q.size();
            long long int start = 0;
            long long int end = 0;
            for (int i = 0; i < qsize; i++) {
                pair<TreeNode*, int> front = q.front();
                long long int index = front.second;
                if (i == 0) {
                    start = front.second;
                }
                if (i == qsize - 1) {
                    end = front.second;
                }
                q.pop();
                if (front.first->left) {
                    q.push({front.first->left, 2 * (index - start) + 1});
                }
                if (front.first->right) {
                    q.push({front.first->right, 2 * (index - start) + 2});
                }
            }
            if (end != 0 && start != 0) {
               
                if(end - start + 1 > maxwidth ){
                    maxwidth= end - start + 1;
                }
            }
        }
        return maxwidth;
    }
};